Bluetooth: Host: Add conn rsp param check

HaavardRei opened the pull request:
Check whether the connection response parameters both with and without ECRED are within the valid ranges from the Bluetooth Core Specification (part 3.A.4 v6.0).

HaavardRei committed:
Bluetooth: Host: Add conn rsp param check

Check whether the connection response parameters both with and without
ECRED are within the valid ranges from the Bluetooth Core Specification
(part 3.A.4 v6.0). Changes validation checks in requests to match the
same pattern.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1929,6 +1929,22 @@ static void le_ecred_conn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
+	if (result == BT_L2CAP_LE_SUCCESS) {
+		if (!L2CAP_LE_CID_IS_DYN(dcid)) {
+			LOG_ERR("dcid 0x%04x is not dynamic", dcid);
+			result = BT_L2CAP_LE_ERR_UNACCEPT_PARAMS;
```
Pretending that the remote peer answered `BT_L2CAP_LE_ERR_UNACCEPT_PARAMS` may not be optimal.

I'm tempted to say that it would be better to disconnect the ACL connection since there has been a protocol error on the L2CAP signalling channel.

If we don't disconnect, then the peer may believe the L2CAP connection is successful, while our side has closed this channel. But you can argue that the peer invoked undefined behavior so we can leave it hanging.

But telling our application or upper layers (`ecred_cb->ecred_conn_rsp`) that the peer responded with something different than it actually did is something we should do very carefully. It can make debugging harder.

Even if we didn't propagate `result` anywhere, clearly this is a just a clever trick to reuse existing code. I think we should avoid clever code if possible. I think it may be a better engineering decision to add new values for `result`. It may infringe on reserved values in the spec, but I would argue that that is a lesser evil. Or we could make `result` be a 32-bit integer instead, giving us room for custom codes. But both of these options may be API breaking to a certain extent. I would like some discussion on this.

Since the correct way to handle this case is not obvious, I would like a comment that explains the choice we make here and why it's fine.

Also, if we are to pretend the peer responded with something, then `BT_L2CAP_LE_ERR_INVALID_PARAMS` would be slightly more appropriate.

Change requested: Please at least use `BT_L2CAP_LE_ERR_INVALID_PARAMS` and write a comment documenting the engineering decisions. (Alternatively let's continue this discussion.)

alwa-nordic replied:
Core 1.E.2.7 allows us to clamp values when that is safe to do:

> If a value is out of range, using the nearest permitted value

I think we should clamp MTU, MPS and credits. Then it's up to the peer to disconnect the channel if we send it an unexpected PDU because of that or if the PDU is bigger than it's willing to accept. This is very safe and tosses the hot potato of disconnecting the channel to the peer.

I think we should accept zero credits in a ECRED connection request even if it breaks spec. It is a well defined state that can be reached later in the connection's life and our app-facing API does not differentiate between ECRED and LECRED, so apps have to tolerate zero credits on connection.

In case of a invalid CID, things get more complicated.

I think the easiest safe thing to do is to disconnect the whole ACL connection.

Alternatively we recover the connection to a good state by sending a disconnect request for the newly created channel. If the remote CID is for a fixed channel, there's no state and we are good. If the remote CID is dynamic but aliases an existing channel, then we should disconnect both channels.

Alternatively to disconnecting both channels, we can give all credits (including the credits given in the connect response) to the existing channels. This is because it's much better for UX to have too many credits rather than leaking them. If credits are leaked, a connection can hang. If we have too many credits the peer must close the connection and we fail fast to a good state.

We can ignore the PDU as like explicitly allowed by Core 1.E.2.7, but I think that can be bad UX?

We can close the channel locally and forget the local CID, like the code does as it is. But the network will not have had a chance to recover to a good state, and if re-use the local CID then we are contributing to the confusion.

HaavardRei replied:
Discussed offline, changed the pattern to:
Receive request with invalid parameters: Send response with corresponding error code as result parameter
Receive response with invalid parameters: Disconnect the ACL connection

HaavardRei committed:
Bluetooth: Host: Use MTU min for l2cap test

The many_conn l2cap test used a MTU lower than the minimum permitted
value. This commit bumps it to the minimum (23).

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1475,7 +1475,11 @@ static void le_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
 	LOG_DBG("psm 0x%02x scid 0x%04x mtu %u mps %u credits %u", psm, scid, mtu, mps, credits);
 
-	if (mtu < L2CAP_LE_MIN_MTU || mps < L2CAP_LE_MIN_MPS) {
+	/* Validate parameters. Requirements are from (Core Spec v6.0, Vol 3.A.4.22). Valid credit
+	 * range is from 0 to UINT16_MAX.
+	 */
+	if (!IN_RANGE(mtu, L2CAP_LE_MIN_MTU, UINT16_MAX) ||
```
The comment mentions a valid credit range (using 0 as the minimal value), but then implements a MTU range check. 

Ideally we'd also define and use a `BT_L2CAP_MAX_MTU` instead of `UINT16_MAX` for the MTU range check.

HaavardRei replied:
The "initial credit" field of the request L2CAP_LE_CREDIT_BASED_CONNECTION_REQ (Core 4.22) and L2CAP_CREDIT_BASED_CONNECTION_REQ (Core 4.25) has respectively the range 0->UINT16_MAX and 1->UINT16_MAX, which is... confusing. 
I added the comment to this parameter check as we don't need to check it as the entire uint16 range is valid (but a reader might wonder why we don't check it). For the case below, the requirements are taken from 4.25 (as mentioned in the comment), where the credit range starts at 1.

Could perhaps clarify instead: "Valid credit range is from 0 to UINT16_MAX, thus the credit parameter does not need validation"?

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1574,8 +1578,12 @@ static void le_ecred_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
-	if (mtu < BT_L2CAP_ECRED_MIN_MTU || mps < BT_L2CAP_ECRED_MIN_MPS) {
-		LOG_ERR("Invalid ecred conn req params. mtu %u mps %u", mtu, mps);
+	/* Validate parameters. Requirements are from Core Spec v6.0, Vol 3.A.4.25. */
+	if (!IN_RANGE(mtu, BT_L2CAP_ECRED_MIN_MTU, UINT16_MAX) ||
+	    !IN_RANGE(mps, BT_L2CAP_ECRED_MIN_MPS, BT_L2CAP_MAX_MPS) ||
+	    !IN_RANGE(credits, 1, UINT16_MAX)) {
```
Here there is a credit range check, but it is using `1` as the minimal value instead of the aforementioned 0.

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1978,10 +1986,17 @@ static void le_ecred_conn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
 			 */
-			if (!dcid) {
+			if (!dcid || !L2CAP_LE_CID_IS_DYN(dcid) ||
+			    !IN_RANGE(mtu, BT_L2CAP_ECRED_MIN_MTU, UINT16_MAX) ||
+			    !IN_RANGE(mps, BT_L2CAP_ECRED_MIN_MPS, BT_L2CAP_MAX_MPS) ||
+			    !IN_RANGE(credits, 1, UINT16_MAX)) {
```
Suggest to add `#define`s for the minimum and maximum credit values too (can be in an internal file if the application doesn't need to care about these values)

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2060,6 +2075,25 @@ static void le_conn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
+			"Terminating connection",
+			dcid, mtu, mps, credits);
+		chan = l2cap_remove_ident(conn, ident);
+		if (chan) {
```
```suggestion
		if (chan != 0) {
```

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1978,10 +1986,17 @@ static void le_ecred_conn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
 			 * established.
 			 */
-			if (!dcid) {
+			if (!dcid || !L2CAP_LE_CID_IS_DYN(dcid) ||
```
```suggestion
			if (dcid == 0U || !L2CAP_LE_CID_IS_DYN(dcid) ||
```

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -111,6 +111,20 @@ extern "C" {
+ *  The maximum MTU for an L2CAP Based Connection. This is the same with or without ECRED. This
+ *  requirement is taken from text in Core 3.A.4.22 and 3.A.4.26 v6.0.
+ */
+#define BT_L2CAP_MAX_MTU UINT16_MAX
```
I have a slight preference to defining this as a literal value (i.e. 65535U or 0xFFFFU), but it's OK as is

HaavardRei replied:
I'll keep it as is unless I get any objections. However, doing a lap and adding the literal type to defines would be nice at a later time!

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1475,7 +1475,11 @@ static void le_conn_req(struct bt_l2cap *l2cap, uint8_t ident,
+	/* Validate parameters. Requirements are from Core Spec v6.0, Vol 3.A.4.22. Valid credit
+	 * range is from 0 to UINT16_MAX, thus no credit validation is needed.
+	 */
+	if (!IN_RANGE(mtu, L2CAP_LE_MIN_MTU, BT_L2CAP_MAX_MTU) ||
```
Previously I've actually seen Coverity complain about similar checks, i.e. checks where the code checks something that's given by the type. So since `BT_L2CAP_MAX_MTU` is UINT16_MAX and `mtu` us a `uint16_t`, we may actually see a Coverity issue for this (similar to checking if an unsigned integer is less than 0). 

You could use the same argument for MAX MTU as you do for credits; it's already implicitly checked by the type itself. 

I'll leave it up to you whether to keep the MAX_MTU check or not
